src/compiler/checker.ts

```diff
@@ -5437,11 +5517,32 @@ module ts {
 
                 return type;
             }
+            
+            function narrowTypeByCallExpression(type: Type, expr: CallExpression, assumeTrue: boolean): Type {
+                // Check that type is not any, assumed result is true.
+                if (type.flags & TypeFlags.Any || !assumeTrue) {
+                    return type;
+                }
+                let callType = checkCallExpression(expr);
+                if (callType.flags & TypeFlags.TypeGuard) {
+                    if (expr.arguments && expr.arguments[(<TypeGuardType>callType).parameterIndex]) {
+                        if (getSymbolAtLocation(expr.arguments[(<TypeGuardType>callType).parameterIndex]) === symbol) {
+                            let narrowedType = (<TypeGuardType>callType).type;
+                            if (isTypeAssignableTo(narrowedType, type)) {
+                               return narrowedType;

```

Comment:
We should only narrow if the type is a subtype of the original type..

you also want to filter in case type is a union type.. 

```ts
 if (targetType) {
                    // Narrow to the target type if it's a subtype of the current type
                    if (isTypeSubtypeOf(narrowedType , type)) {
                        return targetType;
                    }
                    // If the current type is a union type, remove all constituents that aren't subtypes of the target.
                    if (type.flags & TypeFlags.Union) {
                        return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, narrowedType )));
                    }
                }
```